### PR TITLE
Add CLI test for assigning Http proxy to products.

### PR DESCRIPTION
**Pull request by jameerpathan111**

### CODE REVIEW

**Comment by jameerpathan111** [comment]
Test Result:
```
$ pytest tests/foreman/cli/test_product.py -k test_positive_assign_http_proxy_to_products 
=================================================================================== test session starts ===================================================================================
collecting ... 2020-01-28 09:21:44 - conftest - DEBUG - Collected 5 test cases

collected 5 items / 4 deselected / 1 selected                                                                                                                                             

tests/foreman/cli/test_product.py .                                                                                                                                                 [100%]

======================================================================== 1 passed, 4 deselected in 236.51 seconds =========================================================================
```

File: tests/foreman/api/test_repository.py [hunk 1]
```diff
@@ -165,9 +165,6 @@ def test_positive_assign_http_proxy_to_repository(self):
         repo_b.sync({'async': True})
         result = ssh.command('grep -F {} /var/log/messages'.format(proxy_fqdn))
         assert result.return_code == 0
-        # Delete repositories
-        entities.Repository(id=repo_a.id).delete()
-        entities.Repository(id=repo_b.id).delete()
```
Comment (mirekdlugosz) [inline tests/foreman/api/test_repository.py]: Could you explain why this change is made?

File: tests/foreman/cli/test_product.py [hunk 2]
```diff
@@ -246,3 +248,101 @@ def test_product_list_with_default_settings(self):
...
+            'name': gen_string('alpha', 15),
+            'url': http_proxy_url_a,
+            'organization-id': self.org['id'],
+        })
+        http_proxy_b = HttpProxy.create({
+            'name': gen_string('alpha', 15),
+            'url': http_proxy_url_b,
+            'organization-id': self.org['id'],
+        })
+        proxy_fqdn = re.split(r'[:]', http_proxy_b['url'])[1].strip("//")
```
Comment (mirekdlugosz) [inline tests/foreman/cli/test_product.py]: You don't need `re` here, it's simple string split:
```python
http_proxy_b['url'].split(":")[1].strip("//")
```
(Although this way it's not the most readable in the world, perhaps it could be broken down into two lines)

File: tests/foreman/cli/test_product.py [hunk 3]
```diff
@@ -246,3 +248,101 @@ def test_product_list_with_default_settings(self):
...
+        Product.update_proxy({
+            'ids': '{},{}'.format(product_a['id'], product_b['id']),
+            'http-proxy-policy': 'use_selected_http_proxy',
+            'http-proxy-id': http_proxy_b['id']
+        })
+        repo_a1 = Repository.info({'id': repo_a1['id']})
+        repo_a2 = Repository.info({'id': repo_a2['id']})
+        repo_b1 = Repository.info({'id': repo_b1['id']})
+        repo_b2 = Repository.info({'id': repo_b2['id']})
+        # Update following assert statements when BZ#1777713 is fixed.
```
Comment (mirekdlugosz) [inline tests/foreman/cli/test_product.py]: Maybe you want `bz_bug_is_open` here?

File: tests/foreman/api/test_repository.py (hunk 1 above)
Comment (jameerpathan111) [reply tests/foreman/api/test_repository.py]: Because repositories are being synced in background and I cann't delete them at that time, and because of this test is failing.

File: tests/foreman/cli/test_product.py (hunk 3 above)
Comment (jameerpathan111) [reply tests/foreman/cli/test_product.py]: Nope, this works correctly but I am expecting some change to be made in repository info command ones BZ1777713 is fixed. I will update assert statements when the BZ1777713 gets fixed.

File: tests/foreman/cli/test_product.py (hunk 2 above)
Comment (jameerpathan111) [reply tests/foreman/cli/test_product.py]: @mirekdlugosz updated pr.

File: tests/foreman/cli/test_product.py [hunk 4]
```diff
@@ -246,3 +247,101 @@ def test_product_list_with_default_settings(self):
...
+        :expectedresults: HTTP Proxy is assigned to all repos present
+            in Products and sync operation uses assigned http-proxy.
+
+        :CaseImportance: Critical
+        """
+        # create HTTP proxies
+        http_proxy_url_a = '{}:{}'.format(
+            gen_url(scheme='https'), gen_integer(min_value=10, max_value=9999))
+        http_proxy_url_b = '{}:{}'.format(
+            gen_url(scheme='https'), gen_integer(min_value=10, max_value=9999))
```
Comment (vijay8451) [inline tests/foreman/cli/test_product.py]:
```suggestion
            gen_url(scheme='http'), gen_integer(min_value=10, max_value=9999))
```
Above a small suggestion if we could keep URL with `http` as well during proxy create ..

File: tests/foreman/cli/test_product.py [hunk 5]
```diff
@@ -246,3 +247,101 @@ def test_product_list_with_default_settings(self):
...
+        })
+        repo_b1 = make_repository({
+            'product-id': product_b['id'],
+            'url': FAKE_0_YUM_REPO,
+            'http-proxy-policy': 'none'
+        })
+        repo_b2 = make_repository({
+            'product-id': product_b['id'],
+            'url': FAKE_0_YUM_REPO,
+            'http-proxy-policy': 'global_default_http_proxy',
```
Comment (vijay8451) [inline tests/foreman/cli/test_product.py]: By default behaviour is `global_default_http_proxy` when a repo creates .. hence we do not need to specify that here ..

File: tests/foreman/cli/test_product.py [hunk 6]
```diff
@@ -246,3 +247,101 @@ def test_product_list_with_default_settings(self):
...
+            'organization-id': self.org['id'],
+            'async': True
+        })
+        result = ssh.command('grep -F {} /var/log/messages'.format(proxy_fqdn))
+        assert result.return_code == 0
+        # Add http_proxy to products
+        Product.update_proxy({
+            'ids': '{},{}'.format(product_a['id'], product_b['id']),
+            'http-proxy-policy': 'none',
+        })
```
Comment (vijay8451) [inline tests/foreman/cli/test_product.py]: Just a suggestion , If you could add below steps as well after updated the proxy to `none`:
1. Try a product sync 
2. see `/var/log/messages` to check no proxy using now  during sync.

Comment (jameerpathan111) [reply tests/foreman/cli/test_product.py]: @vijay8451 updated the PR.

File: tests/foreman/cli/test_product.py (hunk 5 above)
Comment (jameerpathan111) [reply tests/foreman/cli/test_product.py]: Updated PR.

File: tests/foreman/cli/test_product.py (hunk 4 above)
Comment (jameerpathan111) [reply tests/foreman/cli/test_product.py]: Updated PR.

**Comment by jameerpathan111** [comment]
Updated test result: 
```
$ pytest tests/foreman/cli/test_product.py -k test_positive_assign_http_proxy_to_products
============================= test session starts ==============================
2020-01-29 09:10:11 - conftest - DEBUG - Collected 5 test cases

collected 5 items / 4 deselected / 1 selected

tests/foreman/cli/test_product.py .                                      [100%]

=================== 1 passed, 4 deselected in 304.05 seconds ===================
```

**Comment by vijay8451** [comment]
@jameerpathan111 thanks for addressing the comments.

File: tests/foreman/api/test_repository.py (hunk 1 above)
Comment (mirekdlugosz) [reply tests/foreman/api/test_repository.py]: I wanted to say that perhaps we should update test to wait for repositories to sync, assert on that and then delete, but looking at this particular test purpose, I think you made right call. Syncing is checked elsewhere, we can ignore it here.
